includes/Core/Consent_Mode/Consent_Mode.php

```diff
@@ -80,18 +80,64 @@ public function register() {
 	 * @since n.e.x.t
 	 */
 	protected function render_gtag_consent_snippet() {
-			$consent_defaults = array(
-				'ad_personalization' => 'denied',
-				'ad_storage'         => 'denied',
-				'ad_user_data'       => 'denied',
-				'analytics_storage'  => 'denied',
-				'regions'            => $this->consent_mode_settings->get_regions(),
-			);
-			?>
+		$consent_defaults = array(
+			'ad_personalization' => 'denied',
+			'ad_storage'         => 'denied',
+			'ad_user_data'       => 'denied',
+			'analytics_storage'  => 'denied',
+			'regions'            => $this->consent_mode_settings->get_regions(),
+			'wait_for_update'    => 500, // Allow 500ms for Consent Management Platforms (CMPs) to update the consent status.
+		);
+
+		$consent_category_map = apply_filters(
+			'googlesitekit_consent_category_map',
+			array(
+				'statistics' => array( 'analytics_storage' ),
+				'marketing'  => array( 'ad_storage', 'ad_user_data', 'ad_personalization' ),
+			)
+		);
+		// TODO: We should extract this JS and make it a bit more readable.
+		?>
 <!-- <?php echo esc_html__( 'Google tag (gtag.js) Consent Mode snippet added by Site Kit', 'google-site-kit' ); ?> -->
 <script id='google_gtagjs-js-consent-mode'>
 window.dataLayer = window.dataLayer || [];function gtag(){dataLayer.push(arguments);}
-gtag("consent","default", <?php echo wp_json_encode( $consent_defaults ); ?>);
+gtag("consent","default",<?php echo wp_json_encode( $consent_defaults ); ?>);
+window._googlesitekitConsentCategoryMap = <?php	echo wp_json_encode( $consent_category_map ); ?>;
+document.addEventListener('wp_listen_for_consent_change', function(event) {
+	if (event.detail) {
+		var consentParameters = Object.entries(event.detail).reduce((parameters,[category,status]) => {
+			if (window._googlesitekitConsentCategoryMap[category]) {
+				var mappedStatus = status === 'allow' ? 'granted' : 'denied';
+				window._googlesitekitConsentCategoryMap[category].forEach((parameter) => {
+					parameters[parameter] = mappedStatus;
+				});
+			}
+			return parameters;
+		}, {});
+		if (Object.keys(consentParameters).length) {
+			gtag('consent','update',consentParameters);
+		}
+	}
+});
+function updateGrantedConsent(){
+	if (!(window.wp_consent_type || window.wp_fallback_consent_type)){ return; }
+	if (window._googlesitekitHasUpdatedConsent) {	return; } else { window._googlesitekitHasUpdatedConsent = true; }
+	var consentParameters = Object.entries(window._googlesitekitConsentCategoryMap).reduce((consentParams,[category,parameters]) => {
+		if (window.wp_has_consent && window.wp_has_consent(category)) {
+			parameters.forEach((parameter) => consentParams[parameter] = 'granted');
+		}
+		return consentParams;
+	},{});
+	if (Object.keys(consentParameters).length) {
+		gtag('consent','update',consentParameters);
+	}
+}
+document.addEventListener('wp_consent_type_defined',updateGrantedConsent);
```

Comment:
`wp_consent_type_defined` is an event that may be dispatched by the CMP plugin. For example, here it is being dispatched by Complianz: https://github.com/Really-Simple-Plugins/complianz-gdpr/blob/3beb13674f2d015deb217be2be6e5b2cfdc2a86f/cookiebanner/js/complianz.js#L888-L889

Reply:
It doesn't really seem like it hurts to include this but I'm not sure we should since it's not part of the Consent API, unless I'm missing something? If this is something that should be considered part of the Consent API we should see about having that documented since I don't think we should have plugin-specific integrations as part of this.

Reply:
Hmm that's a good point. I've removed it, for now at least, and let a question about it for @rlankhorst on Slack.